source/Ship.cpp

```diff
@@ -3511,46 +3508,55 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
-	// Cache the maximum speeds of escorts so that the parent can stay
-	// below this speed when the parent wants to keep all escorts together.
-	if(escorts.empty() || escortsVelocity < 0.)
-		escortsVelocity = ship.MaxVelocity();
-	else
-		escortsVelocity = min(escortsVelocity, ship.MaxVelocity());
-	
+	RegisterEscort(ship);
 	escorts.push_back(ship.shared_from_this());
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
+	escortsVelocity = -1.;
+
 	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
-		{
-			escorts.erase(it);
-			break;
-		}
-	
-	// If we remove the slowest escort from a fleet, then the fleet might
-	// be able to speed up.
-	if(ship.MaxVelocity() == escortsVelocity)
+	while(it != escorts.end())
 	{
-		escortsVelocity = -1.;
-
-		it = escorts.begin();
-		for( ; it != escorts.end(); ++it)
-		{
-			auto escort = it->lock();
-			if(escort)
-			{
-				if(escortsVelocity < 0.)
-					escortsVelocity = escort->MaxVelocity();
-				else
-					escortsVelocity = min(escortsVelocity, escort->MaxVelocity());
+		auto escort = it->lock().get();
+		if(escort){
+			if(escort == &ship){
+				escorts.erase(it);
```

Comment:
`it` is invalidated by `erase`, so incrementing after erasure is a bug. You'll need to use `it = escorts.erase(it);` to restore `it` to a valid state (which means you'll need to tweak your iteration strategy here, too).

Reply:
I already changed the loop to use `while` instead of `for` to allow the iterator itself to be replaced and changed the code not to advance the iterator on erase, but I indeed forgot to assign the return-value of the erase action.
Assignment is now also added. Do you see other issues with the loop as it is now?